src/hotspot/share/code/relocInfo.hpp

```diff
@@ -471,24 +473,81 @@ inline relocInfo prefix_relocInfo(int datalen = 0) {
+  // Relocation. We also need the Relocation base class to be at the same
+  // address as the start of the object, e.g. at the address of _relocbuf.
+  // Both of these requirements are checked in the constructors (see
+  // check_reloc_type and check_reloc_placement).  The placement of the base
+  // class subobject isn't guaranteed by C++, since these aren't standard
+  // layout classes, but all supported implementations provide that behavior.
+  // If that changes, we can instead add a Relocation* _reloc member to
+  // capture the result of the placement new, and use that to access the base
+  // subobject.
+  static const size_t _relocbuf_size = 5 * sizeof(void*);
+  union {
+    char _relocbuf[_relocbuf_size];
+    void* alignment_dummy;
+  };
+
+  template<typename Reloc>
+  void check_reloc_type() {
+    static_assert(std::is_base_of<Relocation, Reloc>::value, "not Relocation");
+    static_assert(sizeof(Reloc) <= sizeof(_relocbuf), "_relocbuf too small");
+  }
+
+  // Used to verify that the base class subobject of the object constructed
+  // into _relocbuf is at the same address as the derived object.
+  void check_reloc_placement(const Relocation* reloc) const {
+    assert(static_cast<const void*>(reloc) == _relocbuf, "invariant");
+  }
+
+  // Support for Relocation::copy_into.
+  // reloc should be a most derived object.
+  template<typename Reloc>
+  void copy_into_impl(const Reloc& reloc) {
+    check_reloc_type<Reloc>();
+    Relocation* copy = ::new (_relocbuf) Reloc(reloc);
+    check_reloc_placement(copy);
+  }
+
+  // Support for construct().
+  // ctor is a function object of one argument, the _relocbuf.  It should
+  // construct a relocation object in the given buffer and return a pointer to
+  // that new relocation object.
+  struct Construct {};          // Tag for selecting this constructor.
+  template<typename C> RelocationHolder(Construct, C ctor) {
+    using RelocPtr = decltype(ctor(_relocbuf));
+    static_assert(std::is_pointer<RelocPtr>::value, "expected relocation pointer");
+    using Reloc = std::remove_pointer_t<RelocPtr>;
+    check_reloc_type<Reloc>();
+    Relocation* reloc = ctor(_relocbuf);
+    check_reloc_placement(reloc);
+  }
 
  public:
-  Relocation* reloc() const { return (Relocation*) &_relocbuf[0]; }
+  Relocation* reloc() const { return (Relocation*)_relocbuf; }
   inline relocInfo::relocType type() const;
 
   // Add a constant offset to a relocation.  Helper for class Address.
   RelocationHolder plus(int offset) const;
 
-  inline RelocationHolder();                // initializes type to none
+  // Return a holder containing a relocation of type T, constructed using args.
+  template<typename T, typename... Args>
+  static RelocationHolder construct(const Args&... args) {
+    return RelocationHolder(Construct(),
+                            [&] (void* p) {
+                              return ::new (p) T(args...);
+                            });
+  }
```

Comment:
This would become
```suggestion
  static RelocationHolder construct(const Args&... args) {
    return RelocationHolder(Construct<T>(), args...);
  }
```